### PR TITLE
Added contract check support

### PR DESCRIPTION
**Pull request by shas19**
Added `ContractCheckHandler` which tries to supports Contract Checking. 

Currently, it supports the case when there is only one clause in the contract. The clause of the form in which all the elements of the antecedent are either of "_", "null" or "!null", and the consequent is "!null" is supported. The handler checks and warns under the conditions of the
antecedent if the consequent is "!null" and there is a return statement with `nullable` or `null` expression.

Added unit test

### CODE REVIEW

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java]: Should we have a common utility method for parsing contract annotations rather than having multiple places where we parse the string?

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java]: This `supportedMethods` set is a bit messy and also may be a small memory leak (it never gets cleared).  Maybe instead of doing the check in `onMatchReturn`, we should do a separate traversal inside `onMatchMethod` that finds and checks return statements and reports errors from there?  It is extra traversal, but only for `@Contract` methods.  It might be more efficient than having to look up the enclosing method for every `return` statement.

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java
Comment (msridhar) [inline nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java]: Should this be checking `b != null`?  If this is passing without error right now, looks like something is broken?

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractPositiveCases.java
Comment (msridhar) [inline nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractPositiveCases.java]: This looks off too.  If I'm reading correctly, the contract is referencing the nullability of `b`.

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java]: I am not sure, how this can be done. I might be able to find `ReturnTree` from `MethodTree`. but, not sure how to get the correct `visitorState`.

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java
Comment (shas19) [reply nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java]: `b` is not null according to the contract premises. Hence all the returns are are non-null and hence this is safe. Putting `b != null`, will not change anything as our analysis for contract does not depend on conditionals.

Comment (shas19) [reply nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java:33]: `b` is not null according to the contract premises. Hence all the returns are are non-null and hence this is safe. Putting `b != null`, will not change anything as our analysis for contract does not depend on conditionals.

Comment (msridhar) [reply nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java:33]: On another read, I think this test case is ok.  Sorry for the confusion.

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:69]: Done!

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:106]: Removed the memory leak problem. 

For the alternate method, I am not sure, how this can be done. we might be able to find `ReturnTree` from `MethodTree`. but, not sure how to get the correct `visitorState`.

Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:106]: We can create an appropriate `VisitorState` using the `withPath` method:

https://github.com/google/error-prone/blob/1e2c8ac094d6d8b8c968f722d723800c5a79fb20/check_api/src/main/java/com/google/errorprone/VisitorState.java#L219-L221

I think we should change to doing our own traversal using a `TreePathScanner` inside `onMatchMethod` instead of using `onMatchReturn`.  This way we won't have the overhead of finding the containing method for every `return` statement in the program.

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractUtils.java
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractUtils.java:56]: In various places in this class we have `@Nullable` arguments and then we immediately assert they are non-null.  Why not just get rid of the `@Nullable` and the `assert`?

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractUtils.java
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractUtils.java:51]: This method needs a more descriptive name.  Maybe `reportMatchForContractIssue`?

Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractUtils.java:51]: Please add Javadoc for all methods in this class that don't have any

**Review by msridhar — CHANGES_REQUESTED** [review]
Almost there!  Just a couple more things

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java
Comment (lazaroclapp) [inline nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java:47]: Overall, I feel a few more complex test cases would be illustrative here. What happens with, e.g.

```
  @Contract("!null -> !null")
  public @Nullable T orElse(@Nullable T other) {
    return value != null ? value : other;
  }
```

(from an `Optional` implementation, `value` is a field)

Or:

```
  @Contract("_, _, !null -> !null; _, !null, _ -> !null")
  @Nullable
  public String fooX(
      @Nullable Object a, @Nullable Object b, @Nullable Object c) {
    return b != null ? b : (c != null ? c : a);
  }
```

(based on a simplification of some tricky internal code)

p.s. Although in the later case, I am not 100% we handle multi-predicate contracts yet.

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractPositiveCases.java
Comment (lazaroclapp) [reply nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractPositiveCases.java:32]: Same issue here, the `a != null` check trips me up too. If you want an opaque predicate it, make it about something other than nullability.

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java:72]: Love that this is behind a config flag. So, when you get to my comments on the test cases, we can always figure out the more complicated test case behavior after landing this and seeing if enabling this option internally triggers any false positives for complicated methods with `@Contract` annotations.

File: nullaway/src/main/java/com/uber/nullaway/dataflow/NullnessStoreInitializer.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/dataflow/NullnessStoreInitializer.java:27]: Javadoc?

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractNullnessStoreInitializer.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractNullnessStoreInitializer.java:47]: Do we wish to handle multiple predicate contracts? Do we do so for the standard (non-validating) ContractHandler?

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractNullnessStoreInitializer.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractNullnessStoreInitializer.java:56]: This seems to be ignoring other sources of nullability information, unless I am missing something. So, when analyzing the contract for:

```
@Contract(_, !null -> !null)
@Nullable
public Object foo(Object a, @Nullable Object b) {
   return a;
}
```

the handler will say the contract is not fulfilled, since `a` is not marked `!null` in the antecedent, *but* `a` is unconditionally non-null, so in reality the contract is trivially fulfilled.

Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractNullnessStoreInitializer.java:56]: Basically, it's missing this logic from `CoreNullnessStoreInitializer`:

```
    for (LocalVariableNode param : parameters) {
      Element element = param.getElement();
      Nullness assumed =
          Nullness.hasNullableAnnotation((Symbol) element, config) ? NULLABLE : NONNULL;
      result.setInformation(AccessPath.fromLocal(param), assumed);
    }
    result = handler.onDataflowInitialStore(underlyingAST, parameters, result);
```

(Not sure about how the cascading handlers will interact with the new handler, so that part might be non-trivial, but at least the annotation-sourced nullness should be acknowledged here, I think. Unless I am missing something).

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractUtils.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractUtils.java:29]: Should this also be called by the basic ContractHandler? Can we figure out if there is code duplication between what that handler does to parse this and this new code? (fine to open a follow up issue for that, though)

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractNullnessStoreInitializer.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractNullnessStoreInitializer.java:46]: Why not use the `ContractUtils` parsing logic here too? Maybe `getContractFromAnnotation` should return a list of `Clause` objects and not a string?

File: nullaway/src/main/java/com/uber/nullaway/handlers/BaseNoOpHandler.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/BaseNoOpHandler.java:56]: Why are we making `BaseNoOpHandler` accessible outside of `handlers`. Is the issue that package private doesn't extend to sub-packages?

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractUtils.java
Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractUtils.java:29]: Basically, is [this](https://github.com/uber/NullAway/pull/312/files#diff-24c260e3a0278128d4fc9b1e8926e612R227) now redundant with this method? And, again, might be worth it doing the full parsing and returning an `ImmutableList<ContractClause>` or the like 🙂

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:91]: This answers one of my questions from further ahead in the PR 👍  (I looked at the tests first)

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:106]: Agree with @msridhar here.

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:119]: Not sure this is a great message. Worth: a) showing the contract that is being violated, and b) if possible, explain the violation.

Something like:

```
"Method @Nullable foo(@Nullable a) has @Contract(!null -> !null), but this appears to be violated, as a @Nullable value may be returned even when a is non-null."
```

**Review by lazaroclapp — CHANGES_REQUESTED** [review]
Did a full a pass over this. Overall, it looks great to me! The big two points are the traversal inside the handler (as brought up by @msridhar) and whether or not we are acknowledging annotation-based nullness info when contract checking. Those two are blockers.

For the rest, a lot of them can be done on follow up PRs, including consolidating the `@Contract` parsing code.

File: nullaway/src/main/java/com/uber/nullaway/handlers/BaseNoOpHandler.java
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/BaseNoOpHandler.java:56]: Yes.

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java]: Let me know how it looks now

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java]: Done

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java:72]: Yep!

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java
Comment (shas19) [reply nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java:36]: Currently, we are not using the information from the method signature. I don't know why I didn't do it right now. Let me try if this can be done easily..

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractNullnessStoreInitializer.java
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractNullnessStoreInitializer.java:56]: makes sense. I'll try to use this

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java
Comment (lazaroclapp) [reply nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java:36]: For the first one: I think checking the information from the method signature is important. I don't see us turning this on internally without that, based on the methods I am seeing that use `@Contract` in our codebase :) 

The second is perfectly reasonable given the size of this PR already

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:99]: Wonder how expensive this search is or even the traversal itself. It might be more expensive than traversing up the AST to find the enclosing method, actually, but we can leave the optimization for later, as this is behind an option and we can always profile with/without this handler.

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java
Comment (msridhar) [reply nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java:36]: IIRC we can / should only due heuristic checking of contracts like `null -> false`, as that requires tracking values of boolean variables / expressions.  We could at least catch cases where the contract is `null -> false` but the actual behavior is `null -> true`, where the return is literally `return true`.  I agree this can be added in a follow-up

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:99]: The path used here is incorrect.  It should use `TreePathScanner.getCurrentPath()`:
```suggestion
              state.withPath(getCurrentPath());
```

**Review by msridhar — CHANGES_REQUESTED** [review]
Spotted another small issue

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java]: Don't see a change, actually.

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java
Comment (lazaroclapp) [reply nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java:36]: We can't check 100%, but we can easily do at least everything the stream handler does for `.filter(...)` and boolean returns. It won't cover all cases, but many of this methods are just:

```
@Contract("null -> false")
boolean foo(Object o) {
   // Some code
  if (o == null) {
    // Some error handling
    return false;
  }
  // Some complex logic
}
```

And in that case, we can fully guarantee `null -> false` ;)

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:99]: I checked and the dataflow analysis result is getting cached (at least based on my reading of the code).  This is an extra traversal, but the previous approach was traversing up the AST to find the enclosing method for all `return` statements in the program (to check if the method was one with a `@Contract` annotation).  This should only do work for methods where we've already found a relevant `@Contract` annotation I think.

Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:99]: Right. So the initial one was an approximately `O(M * log N)` cost where `M` is the number of all methods in the analyzed code and `N` the number of nodes in the average method AST (handwaving a lot of things about tree's being balanced and amortizing across method ASTs of vastly different sizes), while this is `O(C * N)` where `C` is the number of methods _with contract annotations_. The key is that we expect `C << M` to an even greater degree than `log N << N`. Correct?

Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:99]: Yes, since method sizes tend to be small, that is my intuition 🙂.  Plus, the previous approach required maintaining some extra instance state to cache which methods were `@Contract` methods, and that set had to be cleared on each top-level class to avoid a leak.  Probably best to avoid such things unless they are really worth it

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java]: I am specifying the contract within the error now and rephrased it a little as you suggested. The snippet above is outdated

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractNullnessStoreInitializer.java
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractNullnessStoreInitializer.java:46]: The current utility methods also report errors and require some extra things like a visitor state. We can add another simple utility method later but I think we can just keep this as it is for now.

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java
Comment (shas19) [reply nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java:36]: The first part is Done 👍🏻

**Review by msridhar — APPROVED** [review]
Looks good!  Before landing please do the renamings below.  (Also @lazaroclapp should approve)

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:111]: This message should still show: method name, as well as note that `@Nullable` is returned when the pre-conditions are satisfied. e.g. 

```
Method [methodName] has @Contract("[contractString]"), but this appears to be violated, as a @Nullable value may be returned when [argument] is `@NonNull`.
```

You can just put "when contract pre-conditions are satisfied" in the case of multiple arguments being relevant, but in the common case of e.g. `_, !null, _ -> !null` it should list the argument's name.

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java
Comment (lazaroclapp) [inline nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractNegativeCases.java:52]: 👍

**Review by lazaroclapp — COMMENTED** [review]
I still think the error message can be improved :)

Also, I made a follow up issue (#424) for the code-duplication issues with parsing the contract. I still feel this should be done in one place only and using a better datastructure, but it can certainly be a follow up PR.

Plus @msridhar 's comments and also this branch should be rebased atop master.

File: nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java:111]: so by argument you mean if we have `return someExpression;` then we use `someExpression` in the message?

Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java]: Let me know how it looks now. I think I was confused because I didn't know that having only one antecedent as `!null` is a very common thing in contracts.

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractPositiveCases.java
Comment (lazaroclapp) [inline nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractPositiveCases.java:45]: I know this was GJF breaking up these lines, but worth formatting them reasonably. For most of this is as easy as deleting the manual space you had added :)

**Review by lazaroclapp — COMMENTED** [review]
I am basically fine with this one as is now, but please fix the tests comments formatting and rebase before this can be ready to merge.

File: nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractPositiveCases.java
Comment (shas19) [reply nullaway/src/test/resources/com/uber/nullaway/testdata/CheckContractPositiveCases.java]: Yeah, this looked weird

**Review by lazaroclapp — APPROVED** [review]
LGTM 🚀

**Pull request merged**
